### DISCUSSION TITLE
modularize cloud storage

### DISCUSSION THREAD

**Discussion by pazos** [koreader/koreader#7267] [Ideas]
This is related with https://github.com/koreader/koreader/discussions/7215 (specially https://github.com/koreader/koreader/discussions/7215#discussioncomment-345094) and a follow up from https://github.com/koreader/koreader/pull/7237.

Mainly I want to get rid of `frontend/apps/cloudstorage` and modularize providers (ie: make them plugins).

So the main `cloudstorage.lua` module gets moved to `frontend` as a dummy implementation, which offers a function to register providers and a function to check if there is *at least* one provider available.

Providers are implemented as plugins. Each provider (dropbox, webdav and ftp) can work without the other ones enabled.

I played a bit with the code, but refactoring it will take a lot of time, so I want to get a bit of feedback first :smile: 

1. `CloudStorage:isReady()` returns true if there's at least one provider enabled. In `frontend/apps/filemanager/filemanagermenu.lua`:

```lua
if CloudStorage:isReady() then
    self.menu_items.cloud_storage = {
        text = _("Cloud storage"),
        callback = function()
            local cloud_storage = CloudStorage:new{}
            UIManager:show(cloud_storage)
            local filemanagerRefresh = function() self.ui:onRefresh() end
            function cloud_storage:onClose()
                filemanagerRefresh()
                UIManager:close(cloud_storage)
            end
        end,
    }
end
```

2. `CloudStorage:registerProvider(id, t)` is used by plugins to register providers and their implementation. A sample ftp provider would be `plugins/provider-ftp.koplugin/main.lua`:

```lua
local CloudStorage = require("cloudstorage")
local FtpProvider = require("ftp-provider")
local WidgetContainer = require("ui/widget/container/widgetcontainer")
local _ = require("gettext")

local provider = WidgetContainer:new{
    name = "provider-ftp",
    is_doc_only = false,
}

function provider:init()
    CloudStorage:registerProvider("ftp", {
        name = _("FTP"),
        config = FtpProvider.config,
        download = FtpProvider.download,
        info = FtpProvider.info,
        list = FtpProvider.list,
    })
end

return provider
```

In that example the `FtpProvider` holds the bits of  FTP implementation  which are now splited in `ftp.lua` and `ftpapi.lua`.

Ideas, suggestions? Is it ok to do that?

**Comment by NiLuJe** [comment]
I approve of the effort, but I can't really tell you much more than that, as I've never used any of those, and I'm unfamiliar with the code in question ;).

**Reply by Frenzie** [reply]
I read through it at the time and probably tested at least some of them but I'm in the same boat.

At a glance it looks like it's simply copying the plugin logic? You could perhaps consider also adding the pcall/not taking down the whole program part of it, but other than that no comments at this point. Also that might be overkill.

**Reply by pazos** [reply]
ftp/webdav/dropbox files don't need major changes. The file `cloudstorage.lua` needs them.

**Comment by poire-z** [comment]
The proposed bits of code looks sane to me.
`CloudStorage:hasProviders()` may be.
Might want to find a way to distinguish "providers" from "real plugins" among the plugins list though.

But please, don't put that kind of discussion in... GitHub Discussions :/ there's no cross ref, and it's the worst to actually having a real discussion :) Best to create a classic Issue "RFC/Discussion" like you've already done numerous times :)

**Reply by pazos** [reply]
Yeah, sorry about that. I will open a proper ticket and hopefully a PR in a few days :+1:
